src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.monoid_algebra.basic
+
+/-!  #  Degree of an `add_monoid_algebra`
+
+Let `R` be a semiring and let `A` be a `semilattice_sup`.
+
+For an element `f : add_monoid_algebra R A`, this file defines
+* `add_monoid_algebra.degree`: the degree taking values in `with_bot A`,
+* `add_monoid_algebra.trailing_degree`: the trailing degree taking values in `with_top A`.
+If the grading type `A` is a linearly ordered additive monoid, then this notions of degree
+coincide with the standard one:
+* the degree is the maximum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊥`, if `f = 0`;
+* the trailing degree is the minimum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊤`, if `f = 0`.
+
+Currently, the only results are
+* `degree_mul_le` -- the degree of a product is at most the sum of the degrees,
+*  `le_trailing_degree_mul` -- the trailing degree of a product is at least the sum of the
+  trailing degrees.
+-/
+
+namespace add_monoid_algebra
+open_locale classical
+
+variables {R A B : Type*} [semiring R]
+
+section degree
+variables [semilattice_sup B] [order_bot B]
+
+/--  Let `R` be a semiring, let `A` be a `semilattice_sup`, and let `f : R[A]` be an
+an element of `add_monoid_algebra R A`.  The degree of `f` takes values in `with_bot A` and
+it is the supremum of the support of `f` or `⊥`, depending on whether `f` is non-zero or not.
+
+If `A` has a linear order, then this notion coincides with the usual one, using the maximum of
+the exponents. -/
+def degree (D : A → B) (f : add_monoid_algebra R A) : B :=
+f.support.sup D
+
+variables [has_add A]
+variables [has_add B] [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+
+lemma degree_mul_le (D : A → B) (Dm : ∀ {a b}, D (a + b) ≤ D a + D b)
+  (f g : add_monoid_algebra R A) :
+  (f * g).degree D ≤ f.degree D + g.degree D :=
+begin
```

Comment:
Of course, maybe there's no point giving it a name here at all, and instead you should leave the specialized names (for the linear order) to `polynomial` and `laurent_polynomial`.

Reply:
I am only hesitant to add results in `monoid_algebra/basic` because it is already a massive file.

So, you are suggesting of proving results about `support.sup D` for `add_monoid_algebras` and then specialize them to results about `degree` in `polynomial/laurent_polynomial`, right?